// @searchspring/snap-preact-components/dist/cjs/components/Organisms/Recommendation/Recommendation.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Recommendation = void 0;
/** @jsx jsx */
var preact_1 = require("preact");
var react_1 = require("@emotion/react");
var classnames_1 = __importDefault(require("classnames"));
var mobx_react_lite_1 = require("mobx-react-lite");
var deepmerge_1 = __importDefault(require("deepmerge"));
var Carousel_1 = require("../../Molecules/Carousel");
var Result_1 = require("../../Molecules/Result");
var utilities_1 = require("../../../utilities");
var providers_1 = require("../../../providers");
var useDisplaySettings_1 = require("../../../hooks/useDisplaySettings");
var ProfileTracker_1 = require("../../Trackers/Recommendation/ProfileTracker");
var ResultTracker_1 = require("../../Trackers/Recommendation/ResultTracker");
var CSS = {
    recommendation: function (_a) {
        var vertical = _a.vertical;
        return (0, react_1.css)({
            height: vertical ? '100%' : undefined,
            '.ss__result__image-wrapper': {
                height: vertical ? '85%' : undefined,
            },
        });
    },
};
exports.Recommendation = (0, mobx_react_lite_1.observer)(function (properties) {
    var _a, _b, _c, _d, _e, _f;
    var globalTheme = (0, providers_1.useTheme)();
    var props = __assign(__assign(__assign({ 
        // default props
        breakpoints: properties.vertical
            ? JSON.parse(JSON.stringify(Carousel_1.defaultVerticalCarouselBreakpoints))
            : JSON.parse(JSON.stringify(Carousel_1.defaultCarouselBreakpoints)), pagination: false, loop: true }, (_a = globalTheme === null || globalTheme === void 0 ? void 0 : globalTheme.components) === null || _a === void 0 ? void 0 : _a.recommendation), properties), (_c = (_b = properties.theme) === null || _b === void 0 ? void 0 : _b.components) === null || _c === void 0 ? void 0 : _c.recommendation);
    var displaySettings = (0, useDisplaySettings_1.useDisplaySettings)(props.breakpoints);
    if (displaySettings && Object.keys(displaySettings).length) {
        var theme = (0, deepmerge_1.default)((props === null || props === void 0 ? void 0 : props.theme) || {}, (displaySettings === null || displaySettings === void 0 ? void 0 : displaySettings.theme) || {}, { arrayMerge: function (destinationArray, sourceArray) { return sourceArray; } });
        props = __assign(__assign(__assign({}, props), displaySettings), { theme: theme });
    }
    var title = props.title, controller = props.controller, children = props.children, breakpoints = props.breakpoints, loop = props.loop, results = props.results, pagination = props.pagination, nextButton = props.nextButton, prevButton = props.prevButton, hideButtons = props.hideButtons, disableStyles = props.disableStyles, style = props.style, className = props.className, vertical = props.vertical, additionalProps = __rest(props, ["title", "controller", "children", "breakpoints", "loop", "results", "pagination", "nextButton", "prevButton", "hideButtons", "disableStyles", "style", "className", "vertical"]);
    if (!controller || controller.type !== 'recommendation') {
        throw new Error("<Recommendation> Component requires 'controller' prop with an instance of RecommendationController");
    }
    var resultsToRender = results || ((_d = controller.store) === null || _d === void 0 ? void 0 : _d.results);
    if (Array.isArray(children) && children.length !== resultsToRender.length) {
        controller.log.error("<Recommendation> Component received invalid number of children. Must match length of 'results' prop or 'controller.store.results'");
        return (0, react_1.jsx)(preact_1.Fragment, null);
    }
    var subProps = {
        carousel: __assign(__assign(__assign({ 
            // default props
            className: 'ss__recommendation__Carousel' }, (_e = globalTheme === null || globalTheme === void 0 ? void 0 : globalTheme.components) === null || _e === void 0 ? void 0 : _e.carousel), (0, utilities_1.defined)({
            disableStyles: disableStyles,
            vertical: vertical,
        })), { 
            // component theme overrides
            theme: props === null || props === void 0 ? void 0 : props.theme }),
        result: __assign(__assign(__assign({ 
            // default props
            className: 'ss__recommendation__result' }, (_f = globalTheme === null || globalTheme === void 0 ? void 0 : globalTheme.components) === null || _f === void 0 ? void 0 : _f.result), (0, utilities_1.defined)({
            disableStyles: disableStyles,
        })), { 
            // component theme overrides
            theme: props === null || props === void 0 ? void 0 : props.theme }),
    };
    var styling = {};
    if (!disableStyles) {
        styling.css = [CSS.recommendation({ vertical: vertical }), style];
    }
    else if (style) {
        styling.css = [style];
    }
    return children || (resultsToRender === null || resultsToRender === void 0 ? void 0 : resultsToRender.length) ? ((0, react_1.jsx)(providers_1.CacheProvider, null,
        (0, react_1.jsx)("div", __assign({}, styling, { className: (0, classnames_1.default)('ss__recommendation', className) }),
            (0, react_1.jsx)(ProfileTracker_1.RecommendationProfileTracker, { controller: controller },
                title && (0, react_1.jsx)("h3", { className: "ss__recommendation__title" }, title),
                (0, react_1.jsx)(Carousel_1.Carousel, __assign({ prevButton: prevButton, nextButton: nextButton, hideButtons: hideButtons, loop: loop, pagination: pagination, breakpoints: breakpoints }, subProps.carousel, additionalProps, displaySettings), Array.isArray(children) && children.length
                    ? children.map(function (child, idx) { return ((0, react_1.jsx)(ResultTracker_1.RecommendationResultTracker, { controller: controller, result: resultsToRender[idx] }, child)); })
                    : resultsToRender.map(function (result) { return ((0, react_1.jsx)(ResultTracker_1.RecommendationResultTracker, { controller: controller, result: result },
                        (0, react_1.jsx)(Result_1.Result, __assign({}, subProps.result, { controller: controller, result: result })))); })))))) : ((0, react_1.jsx)(preact_1.Fragment, null));
});
