// @searchspring/snap-client/dist/esm/Client/apis/Snap.js
import { API } from '.';
export class SnapAPI extends API {
    async postMeta(requestParameters) {
        const headerParameters = {};
        headerParameters['Content-Type'] = 'application/json';
        const response = await this.request({
            path: '/api/v1/meta',
            method: 'POST',
            headers: headerParameters,
            body: requestParameters,
        }, '/api/v1/meta' + JSON.stringify(requestParameters));
        return response;
    }
    async postSearch(requestParameters) {
        const headerParameters = {};
        headerParameters['Content-Type'] = 'application/json';
        const response = await this.request({
            path: '/api/v1/search',
            method: 'POST',
            headers: headerParameters,
            body: requestParameters,
        }, '/api/v1/search' + JSON.stringify(requestParameters));
        return response;
    }
    async postAutocomplete(requestParameters) {
        const headerParameters = {};
        headerParameters['Content-Type'] = 'application/json';
        const response = await this.request({
            path: '/api/v1/autocomplete',
            method: 'POST',
            headers: headerParameters,
            body: requestParameters,
        }, '/api/v1/autocomplete' + JSON.stringify(requestParameters));
        return response;
    }
}
