// @searchspring/snap-preact-components/dist/cjs/components/Atoms/Price/Price.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Price = void 0;
/** @jsx jsx */
var preact_1 = require("preact");
var snap_toolbox_1 = require("@searchspring/snap-toolbox");
var react_1 = require("@emotion/react");
var classnames_1 = __importDefault(require("classnames"));
var providers_1 = require("../../../providers");
var CSS = {
    price: function (_a) {
        var _b;
        var theme = _a.theme;
        return (0, react_1.css)({
            color: (_b = theme === null || theme === void 0 ? void 0 : theme.colors) === null || _b === void 0 ? void 0 : _b.primary,
            '&.ss__price--strike': {
                textDecoration: 'line-through',
                color: 'initial',
            },
        });
    },
};
function Price(properties) {
    var _a, _b, _c;
    var globalTheme = (0, providers_1.useTheme)();
    var theme = __assign(__assign({}, globalTheme), properties.theme);
    var props = __assign(__assign(__assign({ 
        // default props
        symbol: '$', decimalPlaces: 2, padDecimalPlaces: true, thousandsSeparator: ',', decimalSeparator: '.', symbolAfter: false, lineThrough: false }, (_a = globalTheme === null || globalTheme === void 0 ? void 0 : globalTheme.components) === null || _a === void 0 ? void 0 : _a.price), properties), (_c = (_b = properties.theme) === null || _b === void 0 ? void 0 : _b.components) === null || _c === void 0 ? void 0 : _c.price);
    var lineThrough = props.lineThrough, value = props.value, symbol = props.symbol, decimalPlaces = props.decimalPlaces, padDecimalPlaces = props.padDecimalPlaces, thousandsSeparator = props.thousandsSeparator, decimalSeparator = props.decimalSeparator, symbolAfter = props.symbolAfter, raw = props.raw, disableStyles = props.disableStyles, className = props.className, style = props.style;
    var formattedPrice = snap_toolbox_1.filters.currency(value, {
        symbol: symbol,
        decimalPlaces: decimalPlaces,
        padDecimalPlaces: padDecimalPlaces,
        thousandsSeparator: thousandsSeparator,
        decimalSeparator: decimalSeparator,
        symbolAfter: symbolAfter,
    });
    var styling = {};
    if (!disableStyles) {
        styling.css = [CSS.price({ theme: theme }), style];
    }
    else if (style) {
        styling.css = [style];
    }
    return raw ? ((0, react_1.jsx)(preact_1.Fragment, null, formattedPrice)) : ((0, react_1.jsx)(providers_1.CacheProvider, null,
        (0, react_1.jsx)("span", __assign({}, styling, { className: (0, classnames_1.default)('ss__price', { 'ss__price--strike': lineThrough }, className) }), formattedPrice)));
}
exports.Price = Price;
