// @searchspring/snap-preact-components/dist/cjs/components/Molecules/Carousel/Carousel.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Carousel = exports.defaultVerticalCarouselBreakpoints = exports.defaultCarouselBreakpoints = void 0;
/** @jsx jsx */
var preact_1 = require("preact");
var hooks_1 = require("preact/hooks");
var react_1 = require("@emotion/react");
var classnames_1 = __importDefault(require("classnames"));
var mobx_react_lite_1 = require("mobx-react-lite");
var deepmerge_1 = __importDefault(require("deepmerge"));
var core_1 = __importStar(require("swiper/core"));
var Icon_1 = require("../../Atoms/Icon/Icon");
var react_2 = require("swiper/react");
var utilities_1 = require("../../../utilities");
var providers_1 = require("../../../providers");
var useDisplaySettings_1 = require("../../../hooks/useDisplaySettings");
var CSS = {
    carousel: function (_a) {
        var _b, _c;
        var theme = _a.theme, vertical = _a.vertical;
        return (0, react_1.css)({
            display: 'flex',
            maxWidth: '100%',
            maxHeight: vertical ? '100%' : undefined,
            margin: 0,
            padding: 0,
            overflow: 'hidden',
            '.swiper-notification': {
                position: 'absolute',
                left: '100000000000000px',
            },
            '&.ss__carousel-vertical': {
                flexDirection: 'column',
                '.swiper-slide': {
                    /* Center slides vertically */
                    display: 'flex',
                    justifyContent: 'center',
                    alignItems: 'center',
                },
                '.swiper-container': {
                    flexDirection: 'row',
                },
                '.swiper-pagination': {
                    width: 'auto',
                    order: 0,
                    flexDirection: 'column',
                    margin: 0,
                    padding: '10px',
                },
                '.swiper-pagination-bullet': {
                    margin: '4px',
                },
            },
            '.swiper-pagination-bullet-active': {
                background: ((_b = theme === null || theme === void 0 ? void 0 : theme.colors) === null || _b === void 0 ? void 0 : _b.primary) || 'inherit',
            },
            '.ss__carousel__next-wrapper, .ss__carousel__prev-wrapper': {
                display: 'flex',
                justifyContent: 'center',
                alignItems: 'center',
                '&.ss__carousel__next-wrapper--hidden, &.ss__carousel__prev-wrapper--hidden': {
                    display: 'none',
                },
            },
            '.ss__carousel__next, .ss__carousel__prev': {
                padding: '5px',
                cursor: 'pointer',
                lineHeight: 0,
                '&.swiper-button-disabled': {
                    opacity: '0.3',
                    cursor: 'default',
                },
            },
            '.swiper-container': {
                display: 'flex',
                flexDirection: 'column',
                marginLeft: 'auto',
                marginRight: 'auto',
                position: 'relative',
                overflow: 'hidden',
                listStyle: 'none',
                padding: 0,
                zIndex: '1',
            },
            '.swiper-container-vertical': {
                '.swiper-wrapper': {
                    flexDirection: 'column',
                },
            },
            '.swiper-wrapper': {
                order: 0,
                position: 'relative',
                width: '100%',
                height: '100%',
                zIndex: '1',
                display: 'flex',
                transitionProperty: 'transform',
                boxSizing: 'content-box',
            },
            '.swiper-slide': {
                flexShrink: 0,
                width: '100%',
                height: '100%',
                position: 'relative',
                transitionProperty: 'transform',
            },
            '.swiper-pagination': {
                display: 'flex',
                justifyContent: 'center',
                marginTop: '10px',
                width: '100%',
                order: 1,
                transition: '.3s opacity',
            },
            '.swiper-pagination-bullet': {
                width: '8px',
                height: '8px',
                display: 'inline-block',
                borderRadius: '50%',
                background: '#000',
                opacity: '.2',
                cursor: 'pointer',
                margin: '0 4px',
                '&.swiper-pagination-bullet-active': {
                    opacity: '0.8',
                    background: ((_c = theme === null || theme === void 0 ? void 0 : theme.colors) === null || _c === void 0 ? void 0 : _c.primary) || '#000',
                },
            },
            '.swiper-container-pointer-events': {
                touchAction: 'pan-y',
                '&.swiper-container-vertical': {
                    touchAction: 'pan-x',
                },
            },
            '.swiper-slide-invisible-blank': {
                visibility: 'hidden',
            },
        });
    },
};
exports.defaultCarouselBreakpoints = {
    0: {
        slidesPerView: 1,
        slidesPerGroup: 1,
        spaceBetween: 0,
    },
    480: {
        slidesPerView: 2,
        slidesPerGroup: 2,
        spaceBetween: 10,
    },
    768: {
        slidesPerView: 3,
        slidesPerGroup: 3,
        spaceBetween: 10,
    },
    1024: {
        slidesPerView: 4,
        slidesPerGroup: 4,
        spaceBetween: 10,
    },
    1200: {
        slidesPerView: 5,
        slidesPerGroup: 5,
        spaceBetween: 10,
    },
};
exports.defaultVerticalCarouselBreakpoints = {
    0: {
        slidesPerView: 1,
        slidesPerGroup: 1,
        spaceBetween: 0,
    },
};
exports.Carousel = (0, mobx_react_lite_1.observer)(function (properties) {
    var _a, _b, _c, _d;
    var globalTheme = (0, providers_1.useTheme)();
    var theme = __assign(__assign({}, globalTheme), properties.theme);
    var props = __assign(__assign(__assign({ 
        // default props
        breakpoints: properties.vertical
            ? JSON.parse(JSON.stringify(exports.defaultVerticalCarouselBreakpoints))
            : JSON.parse(JSON.stringify(exports.defaultCarouselBreakpoints)), pagination: false, loop: true, autoAdjustSlides: true }, (_a = globalTheme === null || globalTheme === void 0 ? void 0 : globalTheme.components) === null || _a === void 0 ? void 0 : _a.carousel), properties), (_c = (_b = properties.theme) === null || _b === void 0 ? void 0 : _b.components) === null || _c === void 0 ? void 0 : _c.carousel);
    var displaySettings = (0, useDisplaySettings_1.useDisplaySettings)(props.breakpoints);
    if (displaySettings && Object.keys(displaySettings).length) {
        var theme_1 = (0, deepmerge_1.default)((props === null || props === void 0 ? void 0 : props.theme) || {}, (displaySettings === null || displaySettings === void 0 ? void 0 : displaySettings.theme) || {}, { arrayMerge: function (destinationArray, sourceArray) { return sourceArray; } });
        if (props.autoAdjustSlides && props.children.length < displaySettings.slidesPerView) {
            displaySettings.slidesPerView = props.children.length;
            displaySettings.slidesPerGroup = props.children.length;
            displaySettings.loop = false;
        }
        props = __assign(__assign(__assign({}, props), displaySettings), { theme: theme_1 });
    }
    var children = props.children, loop = props.loop, nextButton = props.nextButton, prevButton = props.prevButton, hideButtons = props.hideButtons, vertical = props.vertical, onInit = props.onInit, onNextButtonClick = props.onNextButtonClick, onPrevButtonClick = props.onPrevButtonClick, onClick = props.onClick, disableStyles = props.disableStyles, style = props.style, modules = props.modules, className = props.className, additionalProps = __rest(props, ["children", "loop", "nextButton", "prevButton", "hideButtons", "vertical", "onInit", "onNextButtonClick", "onPrevButtonClick", "onClick", "disableStyles", "style", "modules", "className"]);
    var pagination = props.pagination;
    var subProps = {
        icon: __assign(__assign(__assign({ 
            // default props
            className: 'ss__carousel__icon' }, (_d = globalTheme === null || globalTheme === void 0 ? void 0 : globalTheme.components) === null || _d === void 0 ? void 0 : _d.icon), (0, utilities_1.defined)({
            disableStyles: disableStyles,
        })), { 
            // component theme overrides
            theme: props.theme }),
    };
    var swiperModulesUnfiltered = modules ? [core_1.Navigation, core_1.Pagination, core_1.A11y].concat(modules) : [core_1.Navigation, core_1.Pagination, core_1.A11y];
    //remove any duplicates, in case user passes in Navigation or Pagination
    var swiperModules = swiperModulesUnfiltered.filter(function (module, pos) { return swiperModulesUnfiltered.indexOf(module) === pos; });
    core_1.default.use(swiperModules);
    var navigationPrevRef = (0, hooks_1.useRef)(null);
    var navigationNextRef = (0, hooks_1.useRef)(null);
    var rootComponentRef = (0, hooks_1.useRef)(null);
    var styling = {};
    if (!disableStyles) {
        styling.css = [CSS.carousel({ theme: theme, vertical: vertical }), style];
    }
    else if (style) {
        styling.css = [style];
    }
    if (pagination) {
        if (typeof pagination == 'object') {
            pagination = __assign({ clickable: true }, pagination);
        }
        else {
            pagination = {
                clickable: true,
            };
        }
    }
    return (children === null || children === void 0 ? void 0 : children.length) ? ((0, react_1.jsx)(providers_1.CacheProvider, null,
        (0, react_1.jsx)("div", __assign({ ref: rootComponentRef }, styling, { className: (0, classnames_1.default)('ss__carousel', vertical ? 'ss__carousel-vertical' : '', className) }),
            (0, react_1.jsx)("div", { className: (0, classnames_1.default)('ss__carousel__prev-wrapper', { 'ss__carousel__prev-wrapper--hidden': hideButtons }) },
                (0, react_1.jsx)("div", { className: "ss__carousel__prev", ref: navigationPrevRef, onClick: onPrevButtonClick && (function (e) { return onPrevButtonClick(e); }) }, prevButton || (0, react_1.jsx)(Icon_1.Icon, __assign({ icon: vertical ? 'angle-up' : 'angle-left' }, subProps.icon)))),
            (0, react_1.jsx)(react_2.Swiper, __assign({ centerInsufficientSlides: true, onInit: function (swiper) {
                    //@ts-ignore : someone should refactor this
                    swiper.params.navigation.prevEl = navigationPrevRef.current ? navigationPrevRef.current : undefined;
                    //@ts-ignore : someone should refactor this
                    swiper.params.navigation.nextEl = navigationNextRef.current ? navigationNextRef.current : undefined;
                    if (onInit) {
                        onInit(swiper);
                    }
                }, onClick: function (swiper, e) {
                    onClick && onClick(swiper, e);
                }, direction: vertical ? 'vertical' : 'horizontal', loop: loop, threshold: 7 }, additionalProps, displaySettings, { pagination: pagination }), children.map(function (child) {
                return (0, react_1.jsx)(react_2.SwiperSlide, null, child);
            })),
            (0, react_1.jsx)("div", { className: (0, classnames_1.default)('ss__carousel__next-wrapper', { 'ss__carousel__next-wrapper--hidden': hideButtons }) },
                (0, react_1.jsx)("div", { className: "ss__carousel__next", ref: navigationNextRef, onClick: onNextButtonClick && (function (e) { return onNextButtonClick(e); }) }, nextButton || (0, react_1.jsx)(Icon_1.Icon, __assign({ icon: vertical ? 'angle-down' : 'angle-right' }, subProps.icon))))))) : ((0, react_1.jsx)(preact_1.Fragment, null));
});
