// @searchspring/snap-profiler/dist/cjs/Profiler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Profiler = void 0;
var Profiler = /** @class */ (function () {
    function Profiler(namespace) {
        this.namespace = namespace || '';
        this.profiles = [];
    }
    Profiler.prototype.setNamespace = function (namespace) {
        if (!this.namespace) {
            this.namespace = namespace;
        }
    };
    Profiler.prototype.create = function (_a) {
        var type = _a.type, name = _a.name, context = _a.context;
        if (!name) {
            throw new Error('Profile name is required.');
        }
        var profile = new Profile(this.namespace, { type: type, name: name, context: context });
        this.profiles.push(profile);
        return profile;
    };
    return Profiler;
}());
exports.Profiler = Profiler;
var Profile = /** @class */ (function () {
    function Profile(namespace, _a) {
        var type = _a.type, name = _a.name, context = _a.context;
        this.status = 'pending';
        this.time = {
            date: 0,
            begin: 0,
            end: 0,
            run: 0,
        };
        this.namespace = namespace;
        this.type = type;
        this.name = name;
        this.context = context;
    }
    Profile.prototype.start = function () {
        if (!this.time.begin) {
            this.time.date = Date.now();
            this.time.begin = window.performance.now();
            this.status = 'started';
        }
        return this;
    };
    Profile.prototype.stop = function () {
        if (!this.time.end && this.time.begin) {
            this.time.date = Date.now();
            this.time.end = window.performance.now();
            this.time.run = +(this.time.end - this.time.begin).toFixed(3);
            this.status = 'finished';
        }
        return this;
    };
    return Profile;
}());
