// @searchspring/snap-store-mobx/dist/cjs/Recommendation/RecommendationStore.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.RecommendationStore = void 0;
var mobx_1 = require("mobx");
var AbstractStore_1 = require("../Abstract/AbstractStore");
var Stores_1 = require("../Search/Stores");
var Stores_2 = require("./Stores");
var RecommendationStore = /** @class */ (function (_super) {
    __extends(RecommendationStore, _super);
    function RecommendationStore(config, services) {
        var _this = this;
        var _a;
        _this = _super.call(this, config) || this;
        _this.loaded = false;
        if (typeof services != 'object' || typeof ((_a = services.urlManager) === null || _a === void 0 ? void 0 : _a.subscribe) != 'function') {
            throw new Error("Invalid service 'urlManager' passed to AutocompleteStore. Missing \"subscribe\" function.");
        }
        _this.services = services;
        _this.reset();
        (0, mobx_1.makeObservable)(_this, {
            profile: mobx_1.observable,
            results: mobx_1.observable,
        });
        return _this;
    }
    RecommendationStore.prototype.reset = function () {
        this.update();
    };
    RecommendationStore.prototype.update = function (data) {
        this.error = undefined;
        this.loaded = !!(data === null || data === void 0 ? void 0 : data.profile);
        this.profile = new Stores_2.RecommendationProfileStore(this.services, data);
        this.results = new Stores_1.SearchResultStore(this.config, this.services, data === null || data === void 0 ? void 0 : data.results);
    };
    return RecommendationStore;
}(AbstractStore_1.AbstractStore));
exports.RecommendationStore = RecommendationStore;
