// @searchspring/snap-preact-components/dist/cjs/components/Organisms/Facets/Facets.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Facets = void 0;
/** @jsx jsx */
var preact_1 = require("preact");
var react_1 = require("@emotion/react");
var classnames_1 = __importDefault(require("classnames"));
var mobx_react_lite_1 = require("mobx-react-lite");
var Facet_1 = require("../Facet");
var providers_1 = require("../../../providers");
var utilities_1 = require("../../../utilities");
var CSS = {
    facets: function () { return (0, react_1.css)({}); },
};
exports.Facets = (0, mobx_react_lite_1.observer)(function (properties) {
    var _a, _b, _c, _d, _e, _f;
    var globalTheme = (0, providers_1.useTheme)();
    var props = __assign(__assign(__assign({ 
        // default props
        facets: (_b = (_a = properties.controller) === null || _a === void 0 ? void 0 : _a.store) === null || _b === void 0 ? void 0 : _b.facets }, (_c = globalTheme === null || globalTheme === void 0 ? void 0 : globalTheme.components) === null || _c === void 0 ? void 0 : _c.facets), properties), (_e = (_d = properties.theme) === null || _d === void 0 ? void 0 : _d.components) === null || _e === void 0 ? void 0 : _e.facets);
    var limit = props.limit, disableStyles = props.disableStyles, className = props.className, style = props.style;
    var facets = props.facets;
    if (limit && facets && limit > 0) {
        facets = facets.slice(0, +limit);
    }
    var subProps = {
        facet: __assign(__assign(__assign({ 
            // default props
            className: 'ss__facets__facet' }, (_f = globalTheme === null || globalTheme === void 0 ? void 0 : globalTheme.components) === null || _f === void 0 ? void 0 : _f.facet), (0, utilities_1.defined)({
            disableStyles: disableStyles,
        })), { 
            // component theme overrides
            theme: props.theme }),
    };
    var styling = {};
    if (!disableStyles) {
        styling.css = [CSS.facets(), style];
    }
    else if (style) {
        styling.css = [style];
    }
    return facets && (facets === null || facets === void 0 ? void 0 : facets.length) > 0 ? ((0, react_1.jsx)(providers_1.CacheProvider, null,
        (0, react_1.jsx)("div", __assign({ className: (0, classnames_1.default)('ss__facets', className) }, styling), facets.map(function (facet) { return ((0, react_1.jsx)(Facet_1.Facet, __assign({}, subProps.facet, { facet: facet }))); })))) : ((0, react_1.jsx)(preact_1.Fragment, null));
});
