// @searchspring/snap-store-mobx/dist/cjs/Search/Stores/SearchMerchandisingStore.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SearchMerchandisingStore = exports.ContentType = void 0;
var ContentType;
(function (ContentType) {
    ContentType["HEADER"] = "header";
    ContentType["BANNER"] = "banner";
    ContentType["FOOTER"] = "footer";
    ContentType["LEFT"] = "left";
    ContentType["INLINE"] = "inline";
})(ContentType = exports.ContentType || (exports.ContentType = {}));
var SearchMerchandisingStore = /** @class */ (function () {
    function SearchMerchandisingStore(services, merchData) {
        var _this = this;
        this.redirect = '';
        this.content = {};
        this.campaigns = [];
        if (merchData) {
            this.redirect = merchData.redirect || '';
            if (merchData.content) {
                Object.values(ContentType).forEach(function (type) {
                    if (merchData.content && merchData.content[type]) {
                        _this.content[type] = new Content(merchData.content[type]);
                    }
                });
            }
            if (merchData.campaigns) {
                this.campaigns = merchData.campaigns;
                // if we find a 'landing-page', get landingPage details from merchandising.campaigns
                merchData.campaigns.forEach(function (campaign) {
                    if (campaign.type == 'landing-page') {
                        _this.landingPage = campaign;
                    }
                });
            }
            if (merchData.personalized)
                this.personalized = merchData.personalized;
        }
    }
    return SearchMerchandisingStore;
}());
exports.SearchMerchandisingStore = SearchMerchandisingStore;
var Content = /** @class */ (function (_super) {
    __extends(Content, _super);
    function Content(content) {
        return _super.apply(this, content) || this;
    }
    Object.defineProperty(Content, Symbol.species, {
        get: function () {
            return Array;
        },
        enumerable: false,
        configurable: true
    });
    return Content;
}(Array));
