// @searchspring/snap-store-mobx/dist/cjs/Search/Stores/SearchQueryStore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Query = exports.SearchQueryStore = void 0;
var mobx_1 = require("mobx");
var SearchQueryStore = /** @class */ (function () {
    function SearchQueryStore(services, search) {
        var observables = {};
        if (search === null || search === void 0 ? void 0 : search.query) {
            this.query = new Query(services, search.query);
            observables.query = mobx_1.observable;
        }
        if (search === null || search === void 0 ? void 0 : search.didYouMean) {
            this.didYouMean = new Query(services, search.didYouMean);
            observables.didYouMean = mobx_1.observable;
        }
        if (search === null || search === void 0 ? void 0 : search.originalQuery) {
            this.originalQuery = new Query(services, search.originalQuery);
            observables.originalQuery = mobx_1.observable;
        }
        if (search === null || search === void 0 ? void 0 : search.matchType) {
            this.matchType = search.matchType;
            observables.matchType = mobx_1.observable;
        }
        (0, mobx_1.makeObservable)(this, observables);
    }
    return SearchQueryStore;
}());
exports.SearchQueryStore = SearchQueryStore;
var Query = /** @class */ (function () {
    function Query(services, query) {
        var _a;
        this.string = query;
        this.url = (_a = services === null || services === void 0 ? void 0 : services.urlManager) === null || _a === void 0 ? void 0 : _a.remove('page').remove('filter').set('query', this.string);
        (0, mobx_1.makeObservable)(this, {
            string: mobx_1.observable,
        });
    }
    return Query;
}());
exports.Query = Query;
