// @searchspring/snap-event-manager/dist/cjs/MiddlewareManager.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MiddlewareManager = void 0;
var MiddlewareManager = /** @class */ (function () {
    function MiddlewareManager() {
        this.functions = [];
    }
    MiddlewareManager.prototype.use = function () {
        var _a;
        var func = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            func[_i] = arguments[_i];
        }
        (_a = this.functions).push.apply(_a, func);
    };
    MiddlewareManager.prototype.remove = function (func) {
        var stringyFunc = func.toString();
        this.functions = this.functions.filter(function (fn) {
            return func.name !== fn.name || fn.toString() != stringyFunc;
        });
    };
    MiddlewareManager.prototype.clear = function () {
        this.functions = [];
    };
    MiddlewareManager.prototype.dispatch = function (context) {
        return __awaiter(this, void 0, void 0, function () {
            var cancelling;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, runFunctionsWithAbortWrapper(context || {}, this.functions)];
                    case 1:
                        cancelling = _a.sent();
                        if (cancelling == true) {
                            throw new Error('cancelled');
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    return MiddlewareManager;
}());
exports.MiddlewareManager = MiddlewareManager;
function runFunctionsWithAbortWrapper(context, functions) {
    return __awaiter(this, void 0, void 0, function () {
        var cancelled;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    cancelled = false;
                    return [4 /*yield*/, runFunctions(context, functions, function (proceed) {
                            if (proceed === false)
                                cancelled = true;
                        })];
                case 1:
                    _a.sent();
                    return [2 /*return*/, cancelled];
            }
        });
    });
}
function runFunctions(context, functions, callback) {
    return __awaiter(this, void 0, void 0, function () {
        var func, proceed;
        var _this = this;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!functions.length) {
                        return [2 /*return*/];
                    }
                    func = functions[0];
                    return [4 /*yield*/, func(context, function () { return __awaiter(_this, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, runFunctions(context, functions.slice(1), callback)];
                                    case 1:
                                        _a.sent();
                                        return [2 /*return*/];
                                }
                            });
                        }); })];
                case 1:
                    proceed = _a.sent();
                    callback(proceed);
                    return [2 /*return*/];
            }
        });
    });
}
