// @searchspring/snap-tracker/dist/cjs/TrackEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrackEvent = void 0;
var TrackEvent = /** @class */ (function () {
    function TrackEvent(payload) {
        var _a, _b, _c;
        payload.event = payload.event;
        if (!payload.context || !payload.event) {
            throw 'TrackEvent: object parameter required a valid `context` BeaconContext and `event` ProductClickEvent objects';
        }
        if (!((_a = payload.event) === null || _a === void 0 ? void 0 : _a.intellisuggestData) || !((_b = payload.event) === null || _b === void 0 ? void 0 : _b.intellisuggestSignature)) {
            throw 'TrackEvent: object parameter `event` ProductClickEvent object requires valid intellisuggestData and intellisuggestSignature values. These are the corresponding attributes in the Searchspring API response.';
        }
        this.intellisuggestData = payload.event.intellisuggestData;
        this.intellisuggestSignature = payload.event.intellisuggestSignature;
        this.href = ((_c = payload.event) === null || _c === void 0 ? void 0 : _c.href) || '';
        this.endpoint = "https://".concat(payload.context.website.trackingCode, ".a.searchspring.io/api/track/track.json");
        this.src =
            this.endpoint +
                "?d=".concat(encodeURIComponent(this.intellisuggestData)) +
                "&s=".concat(encodeURIComponent(this.intellisuggestSignature)) +
                "&u=".concat(encodeURIComponent(this.href || ''));
        if (window.document.referrer) {
            this.src += "&r=".concat(encodeURIComponent(window.document.referrer));
        }
        this.img = new Image(1, 1);
        this.img.src = this.src; // setting src immediately invokes a request
    }
    return TrackEvent;
}());
exports.TrackEvent = TrackEvent;
