// @searchspring/snap-store-mobx/dist/cjs/Autocomplete/Stores/AutocompleteQueryStore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutocompleteQueryStore = void 0;
var mobx_1 = require("mobx");
var AutocompleteQueryStore = /** @class */ (function () {
    function AutocompleteQueryStore(services, autocomplete, search, config) {
        var _a;
        var observables = {};
        if (search === null || search === void 0 ? void 0 : search.query) {
            this.query = new Query(services.urlManager, search.query);
            observables.query = mobx_1.observable;
        }
        if (autocomplete === null || autocomplete === void 0 ? void 0 : autocomplete.correctedQuery) {
            if ((_a = config.settings) === null || _a === void 0 ? void 0 : _a.integratedSpellCorrection) {
                this.correctedQuery = new Query(services.urlManager, autocomplete.correctedQuery);
                observables.correctedQuery = mobx_1.observable;
            }
            else if (autocomplete.query) {
                this.originalQuery = new Query(services.urlManager, autocomplete.query);
                observables.originalQuery = mobx_1.observable;
            }
        }
        (0, mobx_1.makeObservable)(this, observables);
    }
    return AutocompleteQueryStore;
}());
exports.AutocompleteQueryStore = AutocompleteQueryStore;
var Query = /** @class */ (function () {
    function Query(urlManager, query) {
        this.string = query;
        this.url = urlManager.set({ query: this.string });
        (0, mobx_1.makeObservable)(this, {
            string: mobx_1.observable,
        });
    }
    return Query;
}());
