// @searchspring/snap-store-mobx/dist/cjs/Autocomplete/Stores/AutocompleteStateStore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Lock = exports.AutocompleteStateStore = void 0;
var mobx_1 = require("mobx");
var AutocompleteStateStore = /** @class */ (function () {
    function AutocompleteStateStore(services) {
        this.focusedInput = undefined;
        this.input = undefined;
        this.locks = {
            terms: new Lock(false),
            facets: new Lock(false),
        };
        this.url = services.urlManager;
        (0, mobx_1.makeObservable)(this, {
            focusedInput: mobx_1.observable,
            locks: mobx_1.observable,
            input: mobx_1.observable,
            reset: mobx_1.action,
        });
    }
    AutocompleteStateStore.prototype.reset = function () {
        this.input = undefined;
        this.locks.terms.reset();
        this.locks.facets.reset();
    };
    return AutocompleteStateStore;
}());
exports.AutocompleteStateStore = AutocompleteStateStore;
var Lock = /** @class */ (function () {
    function Lock(state) {
        if (state === void 0) { state = false; }
        this.state = this.startState = state;
    }
    Lock.prototype.reset = function () {
        this.state = this.startState;
    };
    Object.defineProperty(Lock.prototype, "locked", {
        get: function () {
            return this.state;
        },
        enumerable: false,
        configurable: true
    });
    Lock.prototype.lock = function () {
        this.state = true;
    };
    Lock.prototype.unlock = function () {
        this.state = false;
    };
    return Lock;
}());
exports.Lock = Lock;
