// @searchspring/snap-toolbox/dist/cjs/featureFlags/featureFlags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.featureFlags = exports.getFlags = void 0;
function getFlags(userAgent) {
    if (userAgent === void 0) { userAgent = ''; }
    userAgent = (userAgent || (window.navigator || {}).userAgent || '').toLowerCase();
    var isIE = (function () {
        var ieVersion;
        return function isIE() {
            if (ieVersion === undefined) {
                var version = (userAgent.match(/(msie|trident\/7.0; rv:) ?([0-9]{1,2})\./) || [])[2];
                ieVersion = version ? Number(version) : false;
            }
            return ieVersion;
        };
    })();
    return {
        cors: function () {
            return !isIE() || isIE() >= 10;
        },
        cookies: function () {
            return window.navigator && window.navigator.cookieEnabled && !('doNotTrack' in window.navigator && window.navigator.doNotTrack === '1');
        },
        storage: function () {
            var test = 'ss-test';
            try {
                window.localStorage.setItem(test, test);
                window.localStorage.removeItem(test);
                return true;
            }
            catch (e) {
                return false;
            }
        },
    };
}
exports.getFlags = getFlags;
var flags = getFlags();
exports.featureFlags = {
    cors: flags.cors(),
    cookies: flags.cookies(),
    storage: flags.storage(),
};
