// @searchspring/snap-preact-components/dist/cjs/components/Molecules/FacetListOptions/FacetListOptions.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FacetListOptions = void 0;
/** @jsx jsx */
var preact_1 = require("preact");
var react_1 = require("@emotion/react");
var classnames_1 = __importDefault(require("classnames"));
var mobx_react_lite_1 = require("mobx-react-lite");
var providers_1 = require("../../../providers");
var utilities_1 = require("../../../utilities");
var Checkbox_1 = require("../../Molecules/Checkbox/Checkbox");
var toolbox_1 = require("../../../toolbox");
var CSS = {
    list: function (_a) {
        var _b, _c;
        var theme = _a.theme, hideCheckbox = _a.hideCheckbox;
        return (0, react_1.css)({
            '& .ss__facet-list-options__option': {
                display: 'flex',
                padding: '6px',
                textDecoration: 'none',
                alignItems: 'center',
                '&:hover': {
                    cursor: 'pointer',
                    background: (_b = theme === null || theme === void 0 ? void 0 : theme.colors) === null || _b === void 0 ? void 0 : _b.hover,
                },
                '&.ss__facet-list-options__option--filtered': {
                    fontWeight: 'bold',
                    color: (_c = theme === null || theme === void 0 ? void 0 : theme.colors) === null || _c === void 0 ? void 0 : _c.primary,
                },
                '& .ss__facet-list-options__option__value': {
                    marginLeft: hideCheckbox ? '' : '8px',
                    '& .ss__facet-list-options__option__value__count': {
                        fontSize: '0.8em',
                        marginLeft: '6px',
                    },
                },
            },
        });
    },
};
exports.FacetListOptions = (0, mobx_react_lite_1.observer)(function (properties) {
    var _a, _b, _c, _d;
    var globalTheme = (0, providers_1.useTheme)();
    var theme = __assign(__assign({}, globalTheme), properties.theme);
    var props = __assign(__assign(__assign({}, (_a = globalTheme === null || globalTheme === void 0 ? void 0 : globalTheme.components) === null || _a === void 0 ? void 0 : _a.facetListOptions), properties), (_c = (_b = properties.theme) === null || _b === void 0 ? void 0 : _b.components) === null || _c === void 0 ? void 0 : _c.facetListOptions);
    var values = props.values, hideCheckbox = props.hideCheckbox, hideCount = props.hideCount, onClick = props.onClick, previewOnFocus = props.previewOnFocus, valueProps = props.valueProps, facet = props.facet, disableStyles = props.disableStyles, className = props.className, style = props.style;
    var subProps = {
        checkbox: __assign(__assign(__assign({ 
            // default props
            className: 'ss__facet-list-options__checkbox' }, (_d = globalTheme === null || globalTheme === void 0 ? void 0 : globalTheme.components) === null || _d === void 0 ? void 0 : _d.checkbox), (0, utilities_1.defined)({
            disableStyles: disableStyles,
        })), { 
            // component theme overrides
            theme: props === null || props === void 0 ? void 0 : props.theme }),
    };
    var styling = {};
    if (!disableStyles) {
        styling.css = [CSS.list({ theme: theme, hideCheckbox: hideCheckbox }), style];
    }
    else if (style) {
        styling.css = [style];
    }
    var facetValues = values || (facet === null || facet === void 0 ? void 0 : facet.values);
    return (facetValues === null || facetValues === void 0 ? void 0 : facetValues.length) ? ((0, react_1.jsx)(providers_1.CacheProvider, null,
        (0, react_1.jsx)("div", __assign({}, styling, { className: (0, classnames_1.default)('ss__facet-list-options', className) }), facetValues.map(function (value) {
            var _a, _b;
            return ((0, react_1.jsx)("a", __assign({ className: (0, classnames_1.default)('ss__facet-list-options__option', { 'ss__facet-list-options__option--filtered': value.filtered }), "aria-label": value.filtered
                    ? "remove selected filter ".concat((facet === null || facet === void 0 ? void 0 : facet.label) || '', " - ").concat(value.label)
                    : (facet === null || facet === void 0 ? void 0 : facet.label)
                        ? "filter by ".concat(facet === null || facet === void 0 ? void 0 : facet.label, " - ").concat(value.label)
                        : "filter by ".concat(value.label), href: (_b = (_a = value.url) === null || _a === void 0 ? void 0 : _a.link) === null || _b === void 0 ? void 0 : _b.href }, valueProps, { onClick: function (e) {
                    var _a, _b;
                    (_b = (_a = value.url) === null || _a === void 0 ? void 0 : _a.link) === null || _b === void 0 ? void 0 : _b.onClick(e);
                    onClick && onClick(e);
                } }, (previewOnFocus ? (0, toolbox_1.createHoverProps)(function () { return (value === null || value === void 0 ? void 0 : value.preview) && value.preview(); }) : {})),
                !hideCheckbox && (0, react_1.jsx)(Checkbox_1.Checkbox, __assign({}, subProps.checkbox, { checked: value.filtered, disableA11y: true })),
                (0, react_1.jsx)("span", { className: "ss__facet-list-options__option__value" },
                    value.label,
                    !hideCount && (value === null || value === void 0 ? void 0 : value.count) > 0 && (0, react_1.jsx)("span", { className: "ss__facet-list-options__option__value__count" },
                        "(",
                        value.count,
                        ")"))));
        })))) : ((0, react_1.jsx)(preact_1.Fragment, null));
});
